content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+    ![fig 3: performance metrics cumulative distribution](/images/post/performance-metrics-histogram/fig03_performance_metrics_cumulative_distribution.png)
+
+8. To answer whether the 99th- and 99.9th-percentile latencies are below our target, we simply compare the value in the appropriate bins to the target value. In the illustrated example, the fraction below 1s is 99.8%, and the fraction below 1ms is 24.67%, so it fails the SLA.
+   
+
+### Cost Analysis
+
+- Events are binned in a histogram with pre-determined bin value ranges
+    - e.g. 0-1us, 1us-10us, 10us-100us, 100us-1ms, 1ms-10ms, 10ms-100ms, 100ms-1s, 1s-10s, >10s
+    - Such a histogram is constructed for each second.  
+        __**Space cost: O(|H|) for |H| bins**__  
+        __**Time cost: O(|E|) for |E| events**__ (but this is essentially "free" because it's streaming data.)  
+- The histograms for each second are stored in a list, L.  
+    __**Space cost: O(|L|*|H|)**__  
+- L is sorted by the time. The histograms arrive in order, so this is free.
+- We select the window w from L.  
+    __**Space cost: O(|w|*|H|)**__  
+    __**Time cost: O(|w|)**__ (Note that here |w| is the _number of histograms_, whereas for percentiles |W| is the _number of events_, which is much larger.)
+- We perform a bin-wise addition of the histograms. This means that we take the number of events in a bin and add them across all of the histograms in W, and do the same for each bin. This produces an aggregate histogram (which is just another histogram!).  
+    __**Space cost: O(|H|)**__  
+    __**Time cost: O(|w|*|H|)**__  
+- We then normalize the histogram to obtain population percentage instead of size, by taking the population size associated with each bin and dividing it by the sum of the sizes of all bins to achieve the fraction of the population represented in this bin. Since we care about the fraction of the population whose value _is below a certain value_, we use the sum of the fractions of all bins whose value range is smaller than the current bin, plus the current bin's fraction.  
+    __**Space cost: O(|H|)**__  
+    __**Time cost: O(|H|)**__  
+- We then select the bin whose maximum value is the value we wish a certain percentage of the population to be below and check whether the fraction in that bin is larger than the desired value. If it is, we're good.  
+    __**Time cost: O(1)**__  
+
+__**Total space cost: O(|L|*|H|) + O(|w|*|H|) + O(|H|) * O(|H|)**__  
+__**Total time cost: O(|w|) + O(|w|*|H|) + O(|H|)**__
+
+## Comparing the Two Approaches
+
+The histogram approach is significantly more efficient in both space and time, and it achieves this by use of compression (a histogram is a form of compression since it loses the original data of individual events).
+The efficiency difference becomes important as the total number of values grows; this is true for both network traffic and time as well as storage requirements and query times in the monitoring system.
```

Comment:
I don't follow the `network traffic` comment. The post doesn't explain whether all the raw events are being sent to an external monitoring system or only the percentiles. If it is the latter, then network traffic is worse for the histograms implementation as compared to the sorted list implementation.

Reply:
Good point

Reply:
I added a note at the end of the first approach's cost analysis section that discusses this, which I think provides enough context for the above statement.